File: packages/node/src/client/daemon-lifecycle.ts
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import path from 'node:path';
import { resolveLionscraperJsFromPackageDir } from '../utils/resolve-lionscraper-js.js';
import { daemonHealth } from './daemon-client.js';
import { getDaemonAuthToken, getDaemonHttpBaseUrl } from '../utils/daemon-config.js';
import { logT, portLang } from '../i18n/lang.js';
import { logger } from '../utils/logger.js';

const HEALTH_POLL_MS = 350;
const HEALTH_MAX_WAIT_MS = 20_000;

function resolveLionscraperJsPath(): string {
  const dir = path.dirname(fileURLToPath(import.meta.url));
  return resolveLionscraperJsFromPackageDir(dir);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function waitForHealthy(baseUrl: string, auth: string | undefined): Promise<boolean> {
  const deadline = Date.now() + HEALTH_MAX_WAIT_MS;
  while (Date.now() < deadline) {
    try {
      await daemonHealth(baseUrl, auth);
      return true;
    } catch {
      await sleep(HEALTH_POLL_MS);
    }
  }
  return false;
}

export interface EnsureLocalDaemonResult {
  /** True if this call spawned a new background `lionscraper daemon` (was not already healthy). */
  didSpawn: boolean;
}

/**
 * Ensures the local daemon responds on `getDaemonHttpBaseUrl()` (same `PORT` as this process).
 * If `DAEMON` is `0`, only probes health.
 */
export async function ensureLocalDaemonRunning(): Promise<EnsureLocalDaemonResult> {
  const baseUrl = getDaemonHttpBaseUrl();
  const auth = getDaemonAuthToken();
  const L = portLang();

  try {
    await daemonHealth(baseUrl, auth);
    return { didSpawn: false };
  } catch {
    /* need spawn */
  }

  if (process.env.DAEMON === '0') {
    throw new Error(logT(L, 'daemonUnreachableNoAuto', { baseUrl }));
  }

  const lionscraperJs = resolveLionscraperJsPath();
  const args = ['daemon'];
  if (process.argv.includes('--debug')) {
    args.push('--debug');
  }

  logger.info(logT(L, 'autoDaemonSpawning', { path: lionscraperJs }));

  try {
    const child = spawn(process.execPath, [lionscraperJs, ...args], {
      detached: true,
      stdio: 'ignore',
      windowsHide: true,
      env: process.env,
    });
    child.unref();
  } catch (err) {
    throw new Error(logT(L, 'autoDaemonSpawnFailed', { message: err instanceof Error ? err.message : String(err) }));
  }

  const ok = await waitForHealthy(baseUrl, auth);
  if (!ok) {
    throw new Error(logT(L, 'autoDaemonTimeout', { baseUrl }));
  }
  return { didSpawn: true };
}

File: packages/python/src/lionscraper/client/daemon_lifecycle.py
from __future__ import annotations

import asyncio
import os
import subprocess
import sys

from lionscraper.client.daemon_client import daemon_health
from lionscraper.i18n.lang import log_t, port_lang
from lionscraper.utils.daemon_config import get_daemon_auth_token, get_daemon_http_base_url
from lionscraper.utils.logger import logger

HEALTH_POLL_MS = 350
HEALTH_MAX_WAIT_MS = 20_000


async def _wait_for_healthy(base_url: str, auth: str | None) -> bool:
    deadline = asyncio.get_event_loop().time() + HEALTH_MAX_WAIT_MS / 1000.0
    while asyncio.get_event_loop().time() < deadline:
        try:
            await daemon_health(base_url, auth)
            return True
        except Exception:
            await asyncio.sleep(HEALTH_POLL_MS / 1000.0)
    return False


async def ensure_local_daemon_running() -> dict[str, bool]:
    base_url = get_daemon_http_base_url()
    auth = get_daemon_auth_token()
    L = port_lang()
    try:
        await daemon_health(base_url, auth)
        return {"didSpawn": False}
    except Exception:
        pass

    if os.environ.get("DAEMON", "").strip() == "0":
        raise RuntimeError(log_t(L, "daemonUnreachableNoAuto", {"baseUrl": base_url}))

    args = [sys.executable, "-m", "lionscraper", "daemon"]
    if "--debug" in sys.argv:
        args.append("--debug")

    logger.info(log_t(L, "autoDaemonSpawning", {"path": " ".join(args)}))

    popen_kw: dict = {
        "stdout": subprocess.DEVNULL,
        "stderr": subprocess.DEVNULL,
        "stdin": subprocess.DEVNULL,
    }
    if sys.platform == "win32":
        popen_kw["creationflags"] = subprocess.CREATE_NO_WINDOW | getattr(subprocess, "DETACHED_PROCESS", 0)
    else:
        popen_kw["start_new_session"] = True

    try:
        subprocess.Popen(args, **popen_kw)  # noqa: S603
    except OSError as e:
        raise RuntimeError(log_t(L, "autoDaemonSpawnFailed", {"message": str(e)})) from e

    ok = await _wait_for_healthy(base_url, auth)
    if not ok:
        raise RuntimeError(log_t(L, "autoDaemonTimeout", {"baseUrl": base_url}))
    return {"didSpawn": True}
